core/trino-main/src/main/java/io/trino/sql/planner/TypeAnalyzer.java

```diff
@@ -87,4 +101,73 @@ public static TypeAnalyzer createTestingTypeAnalyzer(PlannerContext plannerConte
                         new TablePropertyManager(),
                         new AnalyzePropertyManager()));
     }
+
+    private static ExpressionAnalyzer createExpressionAnalyzer(Session session,
+                                                               PlannerContext plannerContext,
+                                                               StatementAnalyzerFactory statementAnalyzerFactory,
+                                                               TypeProvider types)
+    {
+        return new ExpressionAnalyzer(plannerContext, new AllowAllAccessControl(), statementAnalyzerFactory, new Analysis(null, ImmutableMap.of(), OTHERS), session, types, WarningCollector.NOOP);
+    }
+
+    private static int computeExpressionsTreeSize(Iterable<? extends Node> expressions)
+    {
+        int size = 0;
+        for (Node expression : expressions) {
+            size += 1;
+            if (!expression.getChildren().isEmpty()) {
+                size += computeExpressionsTreeSize(expression.getChildren());
+            }
+        }
+        return size;
+    }
+
+    private static String getIdFromSession(Session session)
+    {
+        return session.getTransactionId()
+                .map(TransactionId::toString)
+                .orElse(session.getQueryId().getId());
+    }
+
+    private static class CacheIdentityKey
+    {
+        private final String id;
+        private final Iterable<Expression> expressions;
+        private final int treeSize;
+
+        CacheIdentityKey(String id, Iterable<Expression> expressions)
+        {
+            this.id = requireNonNull(id, "id is null");
+            this.expressions = requireNonNull(expressions, "expressions is null");
+            this.treeSize = computeExpressionsTreeSize(expressions);
+        }
+
+        public int getTreeSize()
+        {
+            return treeSize;
+        }
+
+        @Override
+        public boolean equals(Object o)
+        {
+            if (this == o) {
+                return true;
+            }
+
+            if (o == null || getClass() != o.getClass()) {
+                return false;
+            }
+
+            CacheIdentityKey that = (CacheIdentityKey) o;
+
+            return Objects.equals(id, that.id)
+                    && Objects.equals(expressions, that.expressions);
```

Comment:
Expressions should not be compared by equality.
Are you sure it is safe to do so?

Reply:
I strongly opt for caching `Map<RefNode, Type>`, as that's what you actually need here.

The ExpressionAnalyzer was not designed to be created in multiple instances throughout the Planner. It is intended for the Analysis phase, where it provides different kind of information to the Planner via the Analysis object. If you need to analyze expressions later for some reason, there's the `analyzeExpressions` method, and `createConstantAnalyzer` method, depending on your use case. Those methods give you access to the "analyzing" capability of the `ExpressionAnalyzer` while they hide the complexity which is not relevant at that point.